File: C++/1032-sharing.cc
#include <iostream>
#include <array>
#include <iomanip>


int main() {

    std::ios::sync_with_stdio(false);

    int word_a, word_b, n_node;
    std::cin >> word_a >> word_b >> n_node;

    std::array<int, 100000> nexts;
    int addr, next;
    char _;
    for (int i=0; i<n_node; ++i) {
        std::cin >> addr >> _ >> next;
        nexts[addr] = next;
    }
    
    std::array<int, 100000> visited;
    while (word_a != -1) {
        visited[word_a] = 1;
        word_a = nexts[word_a];
    }
    while (word_b != -1) {
        if (visited[word_b]) break;
        word_b = nexts[word_b];
    }
    if (visited[word_b]) {
        std::cout << std::setfill('0')
                  << std::setw(5)
                  << word_b << '\n';
    } else {
        std::cout << -1 << '\n';
    }

    return 0;
}

File: Python/1032-sharing.py
from collections import defaultdict


def common_prefix(word_a, word_b, nexts):
    visited = defaultdict(int)
    while word_a != "-1":
        visited[word_a] = 1
        word_a = nexts[word_a]
    while word_b != "-1":
        if visited[word_b]:
            return word_b
        word_b = nexts[word_b]
    return -1


if __name__ == '__main__':
    word_a, word_b, n_node = input().split()
    nexts = {}
    for _ in range(int(n_node)):
        addr, _, nxt = input().split()
        nexts[addr] = nxt
    print(common_prefix(word_a, word_b, nexts))
